### PR TITLE
Pegasus: restrict social sharing links on /congrats & video player

### PR DESCRIPTION
**Pull request by breville**
Similar to how https://github.com/code-dot-org/code-dot-org/pull/22016 works for dashboard, this change restricts the display of Facebook & Twitter sharing buttons on Pegasus.  It works on the legacy /congrats page and for any page that uses the `share_buttons` partial, which includes the popup video player.

A significant change is an addition of the value `student_y` to the `_user_type` cookie set by the dashboard, signifying a student under 13 years.  It means a student will need to visit a dashboard page after this change goes live before it's set and operates on a pegasus page.  The cookie is necessary so that we have a web cache variant of the pages for students under 13 years.

The code works by hiding the buttons if the `_user_type` indicates a young student, in HAML.  It also shows the (previously hidden, via HAML) social buttons when Facebook and Twitter URLs appear to be reachable.

### CODE REVIEW

**Comment by tanyaparker** [comment]
I believe @wjordan made a change where pages that used the fallback player no longer needed to include `video_player: true`. I think we have a cleanup item on the backlog to remove `video_player: true` from all pages since it's not necessary anymore. Does that affect your assumption that only pages with `video_player: true` use the fallback player? Should we apply to all pegasus pages?

**Comment by breville** [comment]
As it happens, the existence of this flag is somewhat handy, and could be used going forward to make it easy to unify some semi-duplicated video player code.  And as part of this change, I went through and verified that it still exists on all the pages we want this change to apply to, i.e. all the code.org pages that use the popup video player.  So, for now, I rely on the current state of affairs.

If we wanted to apply this fix to even more pegasus pages we could just include this detection behaviour on every pegasus page (and impact every use of the `share_buttons` partial), but it will require a bunch more testing since the scope will be getting bigger.

**Comment by breville** [comment]
After thinking it over, I've prepped another change to this PR which will apply the hiding to all `share_buttons` partials on code.org, with the benefit that it no longer requires `video_player: true`, it no longer requires an additional .js file to be loaded, and it can hide the buttons in HAML until reachability is confirmed.

**Comment by breville** [comment]
This PR is ready for review.  Thanks!

File: dashboard/config/initializers/devise.rb [hunk 1]
```diff
@@ -328,7 +328,15 @@
 
   require 'cookie_helpers'
   Warden::Manager.after_set_user do |user, auth|
-    auth.cookies[environment_specific_cookie_name("_user_type")] = {value: user.teacher? ? "teacher" : "student", domain: :all, httponly: true}
+    user_type =
+      if user.teacher?
+        "teacher"
+      elsif user.under_13?
+        "student_y"
```
Comment (Erin007) [inline dashboard/config/initializers/devise.rb:335]: `student_y` doesn't seem like the most descriptive name - is it short for `young_student`?

**Comment by Erin007** [comment]
Looks mostly good to me! One little nit about naming and a couple of bigger questions:
1.) Do we plan to deprecate legacy congrats? If so is it worth thinking about now since we're in this code? 
2.) Where did we land with having ruby, javascript and inline css in all smashed into one haml file? Is that just the current state of affairs and so we're okay introducing more of it?

**Comment by tanyaparker** [comment]
1) No, we do not currently have a plan for deprecating legacy congrats. It's still the congrats page for 3rd party tutorials because it's more fair and doesn't point to all Code.org things like the new congrats page. If we were to deprecate, we'd have to do significant work infrastructure wise but also to update the new congrats page to be more fair as well for 3rd party tutorials. That's much more work than we have time for right now, but we could look into when we open up Hour of Code work again.

**Comment by breville** [comment]
2. We prefer these to be kept separate.  I made sure new JS in this change comes via the apps pipeline, and tried to leave as much existing code intact as possible rather than refactoring, which is why there's a little more ruby sprinkled into some files alongside existing ruby.  But if I missed an obvious opportunity for better factoring, definitely interested in improving.

File: dashboard/config/initializers/devise.rb (hunk 1 above)
Comment (breville) [reply dashboard/config/initializers/devise.rb:335]: Yes, and agreed that it's not super descriptive.  I deliberated over a few options but didn't want to commit to "young" or a specific age range in the name, staying minimal but still differentiated.  Other ideas?

Comment (Erin007) [reply dashboard/config/initializers/devise.rb:335]: `young_student` is what we use in UI tests to indicate anyone younger than 13, so it seems we could be consistent at least with that.

**Pull request merged**